### PR TITLE
Fixes for `position` with nested `NoopStreams`

### PR DESCRIPTION
**Pull request by nhz2**
With this PR the basic fuzzing tests for `position` while reading finally pass.

### CODE REVIEW

File: src/noop.jl [hunk 1]
```diff
@@ -53,16 +53,18 @@ Note that this method may return a wrong position when
...
+        throw_invalid_mode(mode)
+    elseif mode === :idle
+        return 0
+    elseif has_sharedbuf(stream)
+        return position(stream.stream)
     elseif mode === :write
         return position(stream.stream) + buffersize(stream.buffer1)
-    elseif mode === :read
+    else # read
         return position(stream.stream) - buffersize(stream.buffer1)
```
Comment (nhz2) [inline src/noop.jl:67]: If `buffer1` is shared by `stream.stream` and `stream` this would double count `buffersize`

File: src/noop.jl [hunk 2]
```diff
@@ -53,16 +53,18 @@ Note that this method may return a wrong position when
...
+function Base.position(stream::NoopStream)::Int64
     mode = stream.state.mode
-    if mode === :idle
-        return Int64(0)
+    if !isopen(stream)
+        throw_invalid_mode(mode)
+    elseif mode === :idle
+        return 0
+    elseif has_sharedbuf(stream)
+        return position(stream.stream)
```
Comment (nhz2) [inline src/noop.jl:63]: Here I treat the buffer as being "owned" by the underlying stream even though it is being shared. This is also why I need to add special cases for writing to a `NoopStream` that shares its buffer with the underlying stream. The underlying stream must be notified to change from `:idle` to `:write` mode before its `buffer1` is changed.

File: src/noop.jl [hunk 3]
```diff
@@ -53,16 +53,18 @@ Note that this method may return a wrong position when
 - some data have been inserted by `TranscodingStreams.unread`, or
 - the position of the wrapped stream has been changed outside of this package.
 """
-function Base.position(stream::NoopStream)
+function Base.position(stream::NoopStream)::Int64
```
Comment (mkitti) [inline src/noop.jl:56]: I'm trying to decide if this should be `Int64` or `Int`. The latter varies based on the system word size.

```
$ julia +1.10~x86 -E "typeof(0)"
Int32
```

File: src/noop.jl [hunk 4]
```diff
@@ -53,16 +53,18 @@ Note that this method may return a wrong position when
...
     mode = stream.state.mode
-    if mode === :idle
-        return Int64(0)
+    if !isopen(stream)
+        throw_invalid_mode(mode)
+    elseif mode === :idle
+        return 0
```
Comment (mkitti) [inline src/noop.jl:61]: This would be `Int32` on a 32-bit system.

File: src/noop.jl [hunk 5]
```diff
@@ -97,16 +99,34 @@ function Base.seekend(stream::NoopStream)
     return stream
 end
 
-function Base.unsafe_write(stream::NoopStream, input::Ptr{UInt8}, nbytes::UInt)
+function Base.write(stream::NoopStream, b::UInt8)::Int
+    changemode!(stream, :write)
+    if has_sharedbuf(stream)
+        # directly write data to the underlying stream
+        n = Int(write(stream.stream, b))
```
Comment (mkitti) [inline src/noop.jl:106]: Why do you need the explicit cast to `Int` here?

File: src/noop.jl [hunk 6]
```diff
@@ -97,16 +99,34 @@ function Base.seekend(stream::NoopStream)
...
+    buffer1 = stream.buffer1
+    marginsize(buffer1) > 0 || flushbuffer(stream)
+    return writebyte!(buffer1, b)
+end
+
+function Base.unsafe_write(stream::NoopStream, input::Ptr{UInt8}, nbytes::UInt)::Int
     changemode!(stream, :write)
+    if has_sharedbuf(stream)
+        # directly write data to the underlying stream
+        n = Int(unsafe_write(stream.stream, input, nbytes))
```
Comment (mkitti) [inline src/noop.jl:118]: Why do you need the explicit cast to Int here?

File: src/noop.jl (hunk 3 above)
Comment (nhz2) [reply src/noop.jl:56]: Cool, I didn't know juliaup had 32-bit julia.

This should be `Int64` to support large files.

File: src/noop.jl (hunk 4 above)
Comment (nhz2) [inline src/noop.jl:61]:
```suggestion
        return Int64(0)
```

File: src/noop.jl (hunk 5 above)
Comment (nhz2) [reply src/noop.jl:106]: I need this to fix the `stats` function for `NoopStream`, but that's for a future PR.

Comment (nhz2) [reply src/noop.jl:106]: Right now the cast isn't doing anything.

**Review by mkitti — APPROVED** [review]

**Pull request merged**